exporter/sumologicexporter/compress.go

```diff
@@ -0,0 +1,86 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sumologicexporter
+
+import (
+	"bytes"
+	"compress/flate"
+	"compress/gzip"
+	"io"
+	"io/ioutil"
+)
+
+type compressor struct {
+	format CompressEncodingType
+	writer encoder
+	buf    bytes.Buffer
+}
+
+type encoder interface {
+	io.WriteCloser
+	Reset(dst io.Writer)
+}
+
+// newCompressor takes encoding format and returns compressor struct and error eventually
+func newCompressor(format CompressEncodingType) (compressor, error) {
+	var (
+		writer encoder
+		err    error
+	)
+
+	switch format {
+	case GZIPCompression:
+		writer = gzip.NewWriter(ioutil.Discard)
+	case DeflateCompression:
+		writer, err = flate.NewWriter(ioutil.Discard, flate.BestCompression)
+		if err != nil {
+			return compressor{}, err
+		}
+	default:
+		writer = nil
+	}
+
+	return compressor{
+		format: format,
+		writer: writer,
+	}, nil
+}
+
+// compress takes a reader with uncompressed data and returns
+// a reader with the same data compressed using c.writer
+func (c *compressor) compress(data io.Reader) (io.Reader, error) {
+	if c.writer == nil {
+		return data, nil
+	}
+
+	var dataBytes bytes.Buffer
+	if _, err := dataBytes.ReadFrom(data); err != nil {
+		return nil, err
+	}
+
+	// Reset c.buf to start with empty message
+	c.buf.Reset()
```

Comment:
I believe this is assuming compress can only be called from a single thread. Is it a safe assumption? I figure exporters can process multiple exports concurrently depending on what processors come before, might want to use a pool

Reply:
New instance of `Compressor` is going to be created in every `pushLogsData` to avoid concurrency issues.